[tanguoyang] iP

tanguoyang opened the pull request:
# Dude

> Its like, so cool, bro :heart:

_Dude is your overly relaxed buddy that always seems like he's on something? _

- Read tasks
- Write Tasks
- Delete Tasks
- ~~More weed~~



```bash
add event eventname /from 2001-02-02 /to 2001-03-03
add deadline deadlinename /by 2001-02-02
add todo todoname  
list
delete 0
list
fi
```

**Upcoming features**

- [X] Add task types
- [ ]  Add more tests

[What's a dude?](https://www.urbandictionary.com/define.php?term=dude)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tanguoyang committed:
add greet and exit

tanguoyang committed:
add echo

tanguoyang committed:
add read write to list

tanguoyang committed:
add mark as done

tanguoyang committed:
add deadlines events todos

tanguoyang committed:
update test script

tanguoyang committed:
add error handling

tanguoyang committed:
add delete function

tanguoyang committed:
renamed duke to dude

tanguoyang committed:
wip

tanguoyang committed:
add save to file

tanguoyang committed:
Merge branch 'branch-level-7'

looks good to me

kaiyi27 commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,159 @@
+
+// Base Task class
+abstract class Task {
+    String description;
+    boolean isDone;
```
Maybe could private these variable or limited access of this variable?
I noticed the same issues in other places too.

kaiyi27 commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,159 @@
+public class Dude {
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
+        ArrayList<Task> tasks = new ArrayList<>();
```
Why not putting this as class member instance variable?

kaiyi27 commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,159 @@
+import java.util.Scanner;
+
+// Base Task class
+abstract class Task {
+    String description;
+    boolean isDone;
+
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
+    }
+
+    public void markAsDone() {
+        isDone = true;
+    }
+
+    public void markAsNotDone() {
+        isDone = false;
+    }
+
+    @Override
+    public abstract String toString();
+}
```
Why not splitting different classes into different java class?

kaiyi27 commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,159 @@
+                } else if (input.startsWith("delete ")) {
+                    int index = Integer.parseInt(input.substring(7));
+                    if (index < 0 || index >= tasks.size()) {
+                        throw new IndexOutOfBoundsException("Task number does not exist.");
+                    }
+                    Task removedTask = tasks.remove(index);
+                    System.out.println("Deleted: " + removedTask);
+
+                } else {
+                    String[] parts = input.split(" ", 2);
+                    String command = parts[0];
+                    String taskInfo = parts.length > 1 ? parts[1] : "";
+
+                    switch (command.toLowerCase()) {
+                        case "todo":
+                            tasks.add(new ToDo(taskInfo));
+                            System.out.println("Added ToDo: " + taskInfo);
+                            break;
+                        case "deadline":
+                            if (!taskInfo.contains(" /by ")) {
+                                throw new IllegalArgumentException(
+                                        "Invalid deadline format. Use '/by' to specify the deadline.");
+                            }
+                            String[] deadlineParts = taskInfo.split(" /by ", 2);
+                            tasks.add(new Deadline(deadlineParts[0], deadlineParts[1]));
+                            System.out.println("Added Deadline: " + deadlineParts[0]);
+                            break;
+                        case "event":
+                            if (!taskInfo.contains(" /from ") || !taskInfo.contains(" /to ")) {
+                                throw new IllegalArgumentException(
+                                        "Invalid event format. Use '/from' and '/to' to specify the event times.");
+                            }
+                            String[] eventParts = taskInfo.split(" /from ", 2);
+                            String[] eventTimes = eventParts[1].split(" /to ", 2);
+                            tasks.add(new Event(eventParts[0], eventTimes[0], eventTimes[1]));
+                            System.out.println("Added Event: " + eventParts[0]);
+                            break;
+                        default:
+                            System.out.println("Unknown command");
+                            break;
```
Why not using case for all of the statement to prevent nested if else?
Consider using method to wrap these to make it more readable?

kaiyi27 commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,159 @@
+                }
+            } catch (NumberFormatException e) {
+                System.out.println("Error: Please enter a valid task number.");
+            } catch (IndexOutOfBoundsException e) {
+                System.out.println(e.getMessage());
+            } catch (IllegalArgumentException e) {
+                System.out.println(e.getMessage());
+            } catch (Exception e) {
+                System.out.println("An unexpected error occurred: " + e.getMessage());
```
Maybe can combine these three statement into one as they are having the same behaviour?

kaiyi27 reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where the nesting was too deep. Maybe splitting to different classes makes it more readable. I noted a few minor coding standard violations
too. The main class are getting quite long. Consider splitting into
smaller classes and method if that makes sense.

tanguoyang committed:
add support for different date input and outputs

tanguoyang committed:
Merge branch 'branch-Level-8'
add support for date parsing
.

tanguoyang committed:
add more oop

tanguoyang committed:
add packaging

tanguoyang committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

tanguoyang committed:
add Junit tests

tanguoyang committed:
configure build

tanguoyang committed:
add javadoc comments

tanguoyang committed:
reformat switch cases

tanguoyang committed:
add find

tanguoyang committed:
Merge branch 'branch-Level-9'
Add find function

tanguoyang committed:
Merge branch 'branch-A-CodingStandard'

tanguoyang committed:
Merge branch 'branch-A-JavaDoc'

peterXGD commented on src/test/java/Dude/DudeTest.java:
```diff
@@ -0,0 +1,32 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class DudeTest {
```
I like that you have used the correct naming convention even for test files.

peterXGD commented on src/main/java/Dude/Dude.java:
```diff
@@ -0,0 +1,601 @@
+        String formattedStart = start.format(formatter);
+        String formattedEnd = end.format(formatter);
+        return "[E]" + (isDone ? "[X] " : "[ ] ") + description + " (from: " + formattedStart + " to: " + formattedEnd
+                + ")";
```
I like the fact that you use verb for methods, and nouns for objects, although the word formatted can be a bit confusing.

peterXGD commented on src/main/java/Dude/Dude.java:
```diff
@@ -0,0 +1,601 @@
+            String endDate = event.end.format(formatter);
+            sb.append("E|").append(task.isDone ? "1" : "0").append("|").append(task.description).append("|").append(startDate).append("|").append(endDate);
+        }
+        return sb.toString();
```
I like that you use short name for the more locally used variables and long name for more global variables.

peterXGD commented on src/main/java/Dude/Dude.java:
```diff
@@ -0,0 +1,601 @@
+        System.out.println(message);
+    }
+
+    public void showTaskList(ArrayList<Task> tasks) {
```
I like that other than following naming conventions, you use the keyword "show" for all Ui messages, this makes following using the methods easy with minimal information overloading.

peterXGD commented on src/main/java/Dude/Dude.java:
```diff
@@ -0,0 +1,601 @@
+package Dude;
```
I like that there is a minimal number of nested loops, which shows your good coding ability.

peterXGD reviewed (commented):
LGTM, keep it up 🥇 .

tanguoyang committed:
add gui tutorial

tanguoyang committed:
add GUI to dude

tanguoyang committed:
Merge branch 'add-GUI'
Add gui

tanguoyang committed:
add assertions

tanguoyang committed:
add fromCmd

Add fromCmd for each task to shift command parsing logic to task.

tanguoyang committed:
Merge pull request #2 from tanguoyang/improve-code-readability

add fromCmd

tanguoyang committed:
Merge branch 'master' into add-assertions

tanguoyang committed:
Merge pull request #3 from tanguoyang/add-assertions

add assertions

tanguoyang committed:
add user guide

tanguoyang committed:
rename image

tanguoyang committed:
Merge branch 'branch-add-userguide'
Add user guide

tanguoyang committed:
clean up code

- remove unused methods and attributes

- broke up classes into multiple files to improve readability

tanguoyang committed:
Merge branch 'branch-improve-code-quality'
- Remove unused methods and attributes
- move classes into seperate files for better readability
- add javadoc comments

tanguoyang committed:
Add support for todo duration

add duration to todo
change package Dude to dude

tanguoyang committed:
Merge branch 'branch-add-extension'
add extension

tanguoyang committed:
Update README.md

tanguoyang committed:
Update build.gradle

tanguoyang committed:
Fix bugs

tanguoyang committed:
Fix undo test

tanguoyang committed:
Merge branch 'Fix-ip-review-bugs-branch'

Fix issues brought up in Ip review.

tanguoyang committed:
Fix undone parsing

tanguoyang committed:
Merge branch 'fix-undone-command-branch'
Fix undone not being parsed correctly
